histogram.go

```diff
@@ -0,0 +1,237 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	promproto "github.com/prometheus/client_model/go"
+	"go.uber.org/atomic"
+)
+
+type bucket struct {
+	atomic.Int64
+
+	upper int64 // bucket upper bound, inclusive
+}
+
+type buckets []*bucket
+
+func newBuckets(upperBounds []int64) buckets {
+	bs := make(buckets, 0, len(upperBounds)+1)
+	for _, upper := range upperBounds {
+		bs = append(bs, &bucket{upper: upper})
+	}
+	if upperBounds[len(upperBounds)-1] != math.MaxInt64 {
+		bs = append(bs, &bucket{upper: math.MaxInt64})
+	}
+	return bs
+}
+
+func (bs buckets) get(val int64) *bucket {
+	// Binary search to find the correct bucket for this observation. Bucket
+	// upper bounds are inclusive.
+	i, j := 0, len(bs)
+	for i < j {
+		h := i + (j-i)/2
+		if val > bs[h].upper {
+			i = h + 1
+		} else {
+			j = h
+		}
+	}
+	return bs[i]
+}
+
+// A Histogram approximates a distribution of values. They're both more
+// efficient and easier to aggregate than Prometheus summaries or M3 timers.
+// For a discussion of the tradeoffs between histograms and timers/summaries,
+// see https://prometheus.io/docs/practices/histograms/.
+//
+// All exported methods are safe to use concurrently, and nil *Histograms are
+// valid no-op implementations.
+type Histogram struct {
+	meta       metadata
+	unit       time.Duration
+	bounds     []int64
+	buckets    buckets
+	sum        atomic.Int64 // required by Prometheus
+	labelPairs []*promproto.LabelPair
+}
+
+func newHistogram(m metadata, unit time.Duration, uppers []int64) *Histogram {
+	return &Histogram{
+		buckets:    newBuckets(uppers),
+		meta:       m,
+		unit:       unit,
+		bounds:     uppers,
+		labelPairs: m.MergeLabels(nil /* variable label vals */),
+	}
+}
+
+// Observe finds the correct bucket for the supplied duration and increments
+// its counter.
+func (h *Histogram) Observe(d time.Duration) {
+	if h == nil {
+		return
+	}
+	h.ObserveInt(int64(d / h.unit))
+}
+
+// ObserveInt finds the correct bucket for the supplied integer and increments
+// its counter.
+func (h *Histogram) ObserveInt(n int64) {
+	if h == nil {
+		return
+	}
+	bucket := h.buckets.get(n)
+	bucket.Inc()
+	h.sum.Add(n)
+}
+
+func (h *Histogram) describe() metadata {
+	return h.meta
+}
+
+func (h *Histogram) snapshot() HistogramSnapshot {
+	return HistogramSnapshot{
```

Comment:
should snapshots be nil safe?  If this is called with nil it will panic.

Reply:
I don't think so - I don't think that there's a way for users to trigger that behavior.